unlock-app/src/middlewares/lockMiddleware.js

```diff
@@ -3,123 +3,124 @@
 import React from 'react'
 import { LOCATION_CHANGE } from 'react-router-redux'
 import { CREATE_LOCK, SET_LOCK, WITHDRAW_FROM_LOCK, resetLock } from '../actions/lock'
-import { PURCHASE_KEY, addKey } from '../actions/key'
-import { SET_ACCOUNT, LOAD_ACCOUNT, CREATE_ACCOUNT, setAccount, resetAccountBalance } from '../actions/accounts'
+import { PURCHASE_KEY } from '../actions/key'
+import { SET_ACCOUNT, setAccount } from '../actions/accounts'
 import { setNetwork } from '../actions/network'
 import { setError } from '../actions/error'
 import { SET_PROVIDER } from '../actions/provider'
-import { REFRESH_TRANSACTION, addTransaction, refreshTransaction, updateTransaction, deleteTransaction } from '../actions/transaction'
+import { addTransaction, updateTransaction } from '../actions/transaction'
 
 import Web3Service from '../services/web3Service'
 
 // This middleware listen to redux events and invokes the services APIs.
+// It also listen to events from web3Service and dispatches corresponding actions
+// TODO: consider if on events we should only trigger more actions instead of calling web3Service directly
 export default function lockMiddleware ({ getState, dispatch }) {
 
+  // Buffer of actions waiting for connection
+  const actions = []
+
   const web3Service = new Web3Service()
 
+  /**
+   * When an account was changed, we dispatch the corresponding action
+   * TODO: consider cleaning up state when the account is a different one
+   */
+  web3Service.on('account.changed', (account) => {
+    dispatch(setAccount(account))
+  })
+
+  /**
+   * When a lock was saved, we refresh the balance of its owner
+   */
+  web3Service.on('lock.saved', (lock) => {
+    dispatch(resetLock(lock))
+    web3Service.refreshOrGetAccount(getState().account)
+  })
+
+  /**
+   * The Lock was changed.
+   * Should we get the balance of the lock owner?
+   */
+  web3Service.on('lock.updated', (lock) => {
+    dispatch(resetLock(lock))
+  })
+
+  /**
+   * When a key was saved, we refresh the balance of the lock
+   * and the balance of the account!
+   */
+  web3Service.on('key.saved', (key) => {
+    web3Service.getLock(getState().locks[key.lock])
+    web3Service.refreshOrGetAccount(getState().account)
+  })
+
+  web3Service.on('key.updated', (key) => {
```

Comment:
I think it'd be nice to have comments here as above for consistency. Spell it out.